ludwig/api_annotations.py

```diff
@@ -0,0 +1,185 @@
+        ... def func3(z):
+        ...     return z
+    """
+    if len(args) == 1 and len(kwargs) == 0 and callable(args[0]):
+        return PublicAPI(stability="stable")(args[0])
+
+    if "stability" in kwargs:
+        stability = kwargs["stability"]
+        assert stability in ["stable", "experimental"], stability
+    elif kwargs:
+        raise ValueError(f"Unknown kwargs: {kwargs.keys()}")
+    else:
+        stability = "stable"
+
+    def wrap(obj):
+        if stability == "experimental":
+            message = f"PublicAPI ({stability}): This API is in {stability} " "and may change before becoming stable."
+        else:
+            message = "PublicAPI: This API is stable across Ludwig releases."
+
+        _append_doc(obj, message=message)
+        _mark_annotated(obj)
+        return obj
+
+    return wrap
+
+
+def DeveloperAPI(*args, **kwargs):
+    """Annotation for documenting developer APIs. Developer APIs are lower-level methods explicitly exposed to
+    advanced Ludwig users and library developers. Their interfaces may change across minor Ludwig releases (for
+    e.g., Ludwig 0.6.1 and Ludwig 0.6.2).
+
+    Args:
+        message: a message to help users understand the reason for the deprecation, and provide a migration path.
+    Examples:
+        >>> from api_annotations import DeveloperAPI
+        >>> @DeveloperAPI
+        ... def func(x):
+        ...     return x
+    """
+    if len(args) == 1 and len(kwargs) == 0 and callable(args[0]):
+        return DeveloperAPI()(args[0])
+
+    def wrap(obj):
+        _append_doc(obj, message="DeveloperAPI: This API may change across minor Ludwig releases.")
+        _mark_annotated(obj)
+        return obj
+
+    return wrap
+
+
+def Deprecated(*args, **kwargs):
+    """Annotation for documenting a deprecated API.
+    Deprecated APIs may be removed in future releases of Ludwig (e.g., Ludwig 0.7 to Ludwig 0.8).
+    Args:
+        message: A message to help users understand the reason for the deprecation, and provide a migration path.
+    Examples:
+        >>> from api_annotations import Deprecated
+        >>> @Deprecated
+        ... def func(x):
+        ...     return x
+        >>> @Deprecated(message="g() is deprecated because the API is error prone. Please call h() instead.")
+        ... def g(y):
+        ...     return y
+    """
+    if len(args) == 1 and len(kwargs) == 0 and callable(args[0]):
+        return Deprecated()(args[0])
+
+    message = "**DEPRECATED:** This API is deprecated and may be removed in a future Ludwig release."
+
+    if "message" in kwargs:
+        message += " " + kwargs["message"]
+        del kwargs["message"]
+
+    if kwargs:
+        raise ValueError(f"Unknown kwargs: {kwargs.keys()}")
+
+    def inner(obj):
+        _append_doc(obj, message=message, directive="warning")
+        _mark_annotated(obj)
+        return obj
+
+    return inner
+
+
+def _append_doc(obj, *, message: str, directive: Optional[str] = None) -> str:
```

Comment:
Also I'm not sure why there is a * here in the arg list, I don't see anywhere we pass in more arguments.  how about
```python
def _append_doc(obj, message: str, directive: Optional[str] = None) -> str:
```